cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java

```diff
@@ -0,0 +1,326 @@
+		} while (readLine(is, buffer));
+
+		return refs;
+	}
+
+	/**
+	 * Inspects each element in the given {@code modulePath} to see if it includes a {@code layers.conf} file
+	 * and/or a standard directory structure with child directories {@code system/layers} and, optionally,
+	 * {@code system/add-ons}. If so, the layers identified in {@code layers.conf} are added to the module path
+	 *
+	 * @param modulePath the filesystem locations that make up the standard module path, each of which is to be
+	 *                   checked for the presence of layers and add-ons
+	 *
+	 * @return a new module path, including any layers and add-ons, if found
+	 */
+	public static File[] resolveLayeredModulePath(File... modulePath) {
+		boolean foundLayers = false;
+		List<File> layeredPath = new ArrayList<>();
+
+		for (File file : modulePath) {
+
+			// Always add the root, as the user may place modules directly in it
+
+			layeredPath.add(file);
+
+			LayersConfig layersConfig = _getLayersConfig(file);
+
+			File layersDir = new File(file, layersConfig._getLayersPath());
+
+			if (FileUtil.notExists(layersDir)) {
+				if (layersConfig._isConfigured()) {
+
+					// Bad config from user
+
+					throw new IllegalStateException("No layers directory found at " + layersDir);
+				}
+
+				// else this isn't a root that has layers and add-ons
+
+				continue;
+			}
+
+			boolean validLayers = true;
+			List<File> layerFiles = new ArrayList<>();
+
+			for (String layerName : layersConfig._getLayers()) {
+				File layer = new File(layersDir, layerName);
+
+				if (FileUtil.notExists(layer)) {
+					if (layersConfig._isConfigured()) {
+
+						// Bad config from user
+
+						throw new IllegalStateException(
+							String.format("Cannot find layer %s under directory %s", layerName, layersDir));
+					}
+
+					// else this isn't a standard layers and add-ons structure
+
+					validLayers = false;
+
+					break;
+				}
+
+				loadOverlays(layer, layerFiles);
+			}
+
+			if (validLayers) {
+				foundLayers = true;
+				layeredPath.addAll(layerFiles);
+
+				// Now add-ons
+
+				File[] addOns = new File(file, layersConfig._getAddOnsPath()).listFiles();
+
+				if (addOns != null) {
+					for (File addOn : addOns) {
+						if (addOn.isDirectory()) {
+							loadOverlays(addOn, layeredPath);
+						}
+					}
+				}
+			}
+		}
+
+		if (foundLayers) {
+			return layeredPath.toArray(new File[layeredPath.size()]);
+		}
+
+		return modulePath;
+	}
+
+	public static void safeClose(Closeable closeable) {
+		if (closeable != null) {
+			try {
+				closeable.close();
+			}
+			catch (Throwable ignored) {
+			}
+		}
+	}
+
+	private static LayersConfig _getLayersConfig(File repoRoot) {
+		File layersList = new File(repoRoot, "layers.conf");
+
+		if (FileUtil.notExists(layersList)) {
+			return new LayersConfig();
+		}
+
+		Reader reader = null;
+
+		try (InputStream inputStream = Files.newInputStream(layersList.toPath())) {
+			reader = new InputStreamReader(inputStream, "UTF-8");
```

Comment:
The `InputStreamReader` can be a part of the `try-with-resources` block along with the `InputStream`, so the finally block would not be necessary.

Reply:
ok